### PR TITLE
`paasta status` for Flink on Kubernetes

### PR DESCRIPTION
**Pull request by vkhromov**
```shell
$ paasta status -c kubestage -s compute-infra-test-service -i flink-stream-sql-demo

service: compute-infra-test-service
cluster: kubestage
    instance: flink-stream-sql-demo
    Flink version: 1.6.3
    Jobs: 1 running, 0 finished, 0 failed, 0 cancelled
    10 taskmanagers, 17/20 slots available
    Jobs:
      Job Name                         State       Started
      scribeheartbeat_auditing         RUNNING     2019-02-25 06:44:35 (7 hours ago)

$ paasta status -c kubestage -s compute-infra-test-service -i flink-stream-sql-demo -v

service: compute-infra-test-service
cluster: kubestage
    instance: flink-stream-sql-demo
    Flink version: 1.6.3 54e6cde @ 18.12.2018 @ 02:34:17 UTC
    Jobs: 1 running, 0 finished, 0 failed, 0 cancelled
    10 taskmanagers, 17/20 slots available
    Jobs:
      Job Name                         State       Job ID                           Started
      scribeheartbeat_auditing         RUNNING     eceed55d88a1610fc323efe02ebc9a2e 2019-02-25 06:44:35 (7 hours ago)
```

### CODE REVIEW

**Comment by poros** [comment]
+1 for me, but two things:
1. I am not a maintainer of this repo, so ¯\_(ツ)_/¯
2. The build is failing

I am a bit sorry that you have to special case paasta status (and maybe for paasta logs) for Flink, but I cannot think of anything better which isn't super-complicated at the moment. If we start to have too many customizations for too many workloads, maybe we could go for a plugin system?

**Comment by vkhromov** [comment]
@poros, is the output OK for users of Flink?
Do we need to add anything else into it, or maybe change something?

**Comment by Rob-Johnson** [comment]
I agree with @poros that we need something more pluggable for the future - I want people to be able to have paasta status work without any PR to the paasta codebase. But lets kick the can down the road so we can ship

**Comment by poros** [comment]
> @poros, is the output OK for users of Flink?
> Do we need to add anything else into it, or maybe change something?

@vkhromov , since you were so kind to ask, three things (I am looking at the output in the PR description):
* it would be great to have the flink dashboard link for the instance there
* perhaps the version of flink-supervisor installed when called with `-v` could be useful for bulk updates (even if it's implied by the docker image hash)
* in paasta, when `-v` is passed, you have the list of containers. Can it be done here for k8s? This is unrelated to Flink, I suppose

**Comment by poros** [comment]
Looking at mesos/marathon infos in paasta status, I would say that the interesting info are the container:host mapping, the health of each container, the RAM and CPU and how long the container has been up:
```
Marathon:   Healthy - up with (4/4) instances. Status: Running
      Marathon dashboard: http://marathon1.paasta-norcal-stagef.yelp/ui/#/apps/%2Fkew.main.git94a0675d.configb8e10538
        4 running, 4 healthy, 0 staged out of 4
        App created: 2019-01-23 16:56:26.503000 (a month ago)
        Status: Running
        Tasks:
          Mesos Task ID                         Host deployed to                     Deployed at what localtime    Health
          a3d6d522-38a7-11e9-8ed5-0242563309be  hostA:31429   2019-02-24T18:47 (a day ago)  Healthy
          b8cae843-38a7-11e9-8ed5-0242563309be  hostB:31547    2019-02-24T18:47 (a day ago)  Healthy
          b8cb5d74-38a7-11e9-8ed5-0242563309be  hostC:31224  2019-02-24T18:47 (a day ago)  Healthy
          d2755785-38a7-11e9-8ed5-0242563309be  hostD:31059  2019-02-24T18:48 (a day ago)  Healthy
    Mesos:      Healthy - (4/4) tasks in the TASK_RUNNING state.
      Running Tasks:
        Mesos Task ID                         Host deployed to               Ram          CPU   Deployed at what localtime
        d2755785-38a7-11e9-8ed5-0242563309be  hostD  1583/4032MB  0.4%  2019-02-24T18:48 (a day ago)
        b8cae843-38a7-11e9-8ed5-0242563309be  hostB    1560/4032MB  0.2%  2019-02-24T18:47 (a day ago)
        a3d6d522-38a7-11e9-8ed5-0242563309be  hostA   1578/4032MB  0.3%  2019-02-24T18:47 (a day ago)
        b8cb5d74-38a7-11e9-8ed5-0242563309be  hostC  1572/4032MB  0.2%  2019-02-24T18:47 (a day ago)
      Non-Running Tasks
        Mesos Task ID  Host deployed to  Deployed at what localtime  Status
    Smartstack:
      Haproxy Service Name: kew.main
      Backends:
          Name                                                LastCheck      LastChange     Status
        uswest1-stagef - Healthy - in haproxy with (4/4) total backends UP in this namespace.
          10.112.53.126:31429:hostA    L7OK/0 in 0ms  5 minutes ago  UP
          10.112.38.65:31547:hostB      L7OK/0 in 0ms  5 minutes ago  UP
          10.112.165.144:31224:hostC  L7OK/0 in 0ms  5 minutes ago  UP
          10.112.163.112:31059:1hostD  L7OK/0 in 0ms  5 minutes ago  UP
          10.40.11.166:31413:hostA       / in ms        6 minutes ago  MAINT
          10.40.23.150:31196:hostB        / in ms        6 minutes ago  MAINT
          10.40.11.47:31682:hostC          / in ms        6 minutes ago  MAINT
          10.40.23.150:31163:hostD       / in ms        6 minutes ago  MAINT
```

Anyway, as for now we have the workaround of pulling these kind of info from kubectl, so this could wait, I would say:
```
kubectl get pods  --namespace paasta-flinkclusters
NAME                                                READY   STATUS    RESTARTS   AGE
flink-cluster-operator-f55489745-wk4wf              1/1     Running   0          5h17m
flink-stream-sql-demo-jobmanager-6c6b4fdd74-h5w67   2/2     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-49tw9   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-5htgh   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-6x7td   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-8vncm   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-fg78q   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-kj5gt   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-nlmnj   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-r8ws2   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-rzfqw   1/1     Running   0          5h17m
flink-stream-sql-demo-taskmanager-fbf7ffbfc-s7zdg   1/1     Running   0          5h17m
```

File: paasta_tools/api/views/instance.py [hunk 1]
```diff
@@ -240,21 +258,29 @@ def instance_status(request):
...
     except Exception:
         error_message = traceback.format_exc()
         raise ApiFailure(error_message, 500)
 
-    version = get_deployment_version(actual_deployments, settings.cluster, instance)
-    # exit if the deployment key is not found
-    if not version:
-        error_message = 'deployment key %s not found' % '.'.join([settings.cluster, instance])
-        raise ApiFailure(error_message, 404)
+    if instance_type != 'flinkcluster':
```
Comment (keymone) [inline paasta_tools/api/views/instance.py]: i'm bothered by growing amount of `if`s on instance_type. we  should refactor the whole thing into dynamic dispatch on  instance_type value, merging the result into `instance_state`, something like:

    INSTANCE_STATUS_HANDLERS = {'flinkcluster': ..., ...}
    def instance_status():
      instance_type = ...
      instance_status.update(INSTANCE_STATUS_HANDLERS[instance_type](service, instance, verbose, ...))
      return instance_status

File: paasta_tools/cli/cmds/status.py [hunk 2]
```diff
@@ -327,6 +337,60 @@ def status_kubernetes_job_human(
...
+        output.append(fmt.format(
+            job_id=job_id,
+            job_name=job['name'].split('.', 2)[2],
+            state=job['state'],
+            start_time=f'{str(start_time)} ({humanize.naturaltime(start_time)})',
+        ))
+        if job_id in status.exceptions:
+            exceptions = status.exceptions[job_id]
+            root_exception = exceptions['root-exception']
+            if root_exception is not None:
```
Comment (keymone) [inline paasta_tools/cli/cmds/status.py]: can there be not-None `exceptions` without root-exception?

Comment (vkhromov) [reply paasta_tools/cli/cmds/status.py]: Nope
There is always `"root-exception"` key in `exceptions`, please see the flink operator src for details.

File: paasta_tools/api/views/instance.py (hunk 1 above)
Comment (vkhromov) [reply paasta_tools/api/views/instance.py]: I'm totally agree with that, we need to do that.
Let's do this refactoring when we'll have a stable working code for several operators.

File: paasta_tools/cli/cmds/status.py (hunk 2 above)
Comment (keymone) [reply paasta_tools/cli/cmds/status.py]: then we  should probably change this `if` to something like

    if root_eception  is None:
      log.error()
      return

Comment (vkhromov) [reply paasta_tools/cli/cmds/status.py]: There are two cases for the `exceptions` field:
1. a job without exception:
```yaml
  exceptions:
    2f9a6671abd1894d794262ef14a1bf3e:
      root-exception: null
      timestamp: null
```
2. a job with an exception:
```yaml
  exceptions:
    2f9a6671abd1894d794262ef14a1bf3e:
      root-exception: "somethingsomething"
      timestamp: 1551164324119
```
Why we should call `log.error()` and stop processing if `root_exception` is `None` (or even if it isn't `None`)?

Comment (keymone) [reply paasta_tools/cli/cmds/status.py]: i  guess i misunderstood you when you said it's impossible to have `exceptions`  field without `root-exception`. for me having `root-exception: nil` means not having `root-exception` :)

is there a reason we include an entry with empty `root-exception` at all?

Comment (vkhromov) [reply paasta_tools/cli/cmds/status.py]: This is Fink , not us.
See https://ci.apache.org/projects/flink/flink-docs-stable/monitoring/rest_api.html

Comment (keymone) [reply paasta_tools/cli/cmds/status.py]: maybe doing omit-empty on exception  related  fields would make  golang serializer skip empty  maps? anyway, not related to this PR.

Comment (vkhromov) [reply paasta_tools/cli/cmds/status.py]: At the moment values in `Status` have the same types as corresponding Flink REST API responses.
IMO we better continue using those types.

**Pull request merged**